Tests/NIOKitTests/ConnectionPoolTests.swift

```diff
@@ -67,6 +69,52 @@ public final class ConnectionPoolTests: XCTestCase {
         XCTAssert(a === c)
     }
     
+    
+    func testConnectError() throws {
+        let db = ErrorDatabase()
+        let pool = ConnectionPool(config: .init(maxConnections: 1), source: db)
+        defer { try! pool.close().wait() }
+        do {
+            _ = try pool.requestConnection().wait()
+            XCTFail("should not have created connection")
+        } catch _ as ErrorDatabase.Error {
+            // pass
+        }
+        
+        // test that we can still make another request even after a failed request
+        do {
+            _ = try pool.requestConnection().wait()
+            XCTFail("should not have created connection")
+        } catch _ as ErrorDatabase.Error {
+            // pass
+        }
+    }
+    
+    func testPoolClose() throws {
+        let foo = FooDatabase()
+        let pool = ConnectionPool(config: .init(maxConnections: 1), source: foo)
+        let _ = try pool.requestConnection().wait()
+        let b = pool.requestConnection()
+        try pool.close().wait()
+        let c = pool.requestConnection()
+        
+        // check that waiters are failed
+        do {
+            _ = try b.wait()
+            XCTFail("should not have created connection")
+        } catch _ as ConnectionPoolError {
+            // pass
```

Comment:
To future proof, I think we should check for the `.closed` error here, since there might be more/different errors in the future.